pkg/xds/translate.go

```diff
@@ -40,8 +41,186 @@ const (
+							Action: &route.Route_Route{
+								Route: &route.RouteAction{
+									ClusterSpecifier: &route.RouteAction_Cluster{Cluster: cluster},
+								},
+							},
+						},
+					}},
+				})
+			}
+		}
+	}
+
+	return out
+}
+
+// MatchServiceHosts splits the virtual service hosts into services and literal hosts
+func MatchServiceHosts(in model.Config, serviceByName ServiceByName) ([]string, []*model.Service) {
+	rule := in.Spec.(*routingv2.RouteRule)
+	hosts := make([]string, 0)
+	services := make([]*model.Service, 0)
+	for _, host := range rule.Hosts {
+		if svc := serviceByName(host, in.ConfigMeta.Namespace); svc != nil {
+			services = append(services, svc)
+		} else {
+			hosts = append(hosts, host)
+		}
+	}
+	return hosts, services
+}
+
+// TranslateVirtualHost creates virtual hosts corresponding to a virtual service.
+func TranslateVirtualHost(in model.Config, serviceByName ServiceByName, subsetSelector SubsetSelector) []GuardedHost {
+	hosts, services := MatchServiceHosts(in, serviceByName)
+	serviceByPort := make(map[int][]*model.Service)
+	for _, svc := range services {
+		for _, port := range svc.Ports {
+			if port.Protocol.IsHTTP() {
+				serviceByPort[port.Port] = append(serviceByPort[port.Port], svc)
+			}
+		}
+	}
+
+	// if no services matched, then we have no port information -- default to 80 for now
+	// TODO: use match condition ports
+	if len(serviceByPort) == 0 {
+		serviceByPort[80] = nil
+	}
+
+	out := make([]GuardedHost, len(serviceByPort))
+	for port, services := range serviceByPort {
+		clusterNaming := TranslateDestination(serviceByName, subsetSelector, in.ConfigMeta.Namespace, port)
+		routes := TranslateRoutes(in, clusterNaming)
+		out = append(out, GuardedHost{
+			Port:     port,
+			Services: services,
+			Hosts:    hosts,
+			Routes:   routes,
+		})
+	}
+
+	return out
+}
+
+// TranslateDestination produces a cluster naming function using the config context.
+func TranslateDestination(
+	serviceByName ServiceByName,
+	subsetSelector SubsetSelector,
+	contextNamespace string,
+	defaultPort int) ClusterNaming {
+	return func(destination *routingv2.Destination) string {
+		// detect if it is a service
+		svc := serviceByName(destination.Name, contextNamespace)
+
+		// TODO: create clusters for non-service hostnames/IPs
+		if svc == nil {
+			return UnresolvedCluster
+		}
+
+		// default port uses port number
+		svcPort, _ := svc.Ports.GetByPort(defaultPort)
+		if destination.Port != nil {
+			switch selector := destination.Port.Port.(type) {
+			case *routingv2.PortSelector_Name:
+				svcPort, _ = svc.Ports.Get(selector.Name)
+			case *routingv2.PortSelector_Number:
+				svcPort, _ = svc.Ports.GetByPort(int(selector.Number))
+			}
+		}
+
+		if svcPort == nil {
+			return UnresolvedCluster
+		}
+
+		// use subsets if it is a service
+		labels := subsetSelector(svc, destination.Subset)
+		key := svc.Key(svcPort, labels)
+		return v1.TruncateClusterName(v1.OutboundClusterPrefix + key)
```

Comment:
The return value of this function should be the cluster name and a metadata struct like what Envoy has. This will allow us to fully fill out the weighted cluster entry (instead of just using cluster name). This has no pertinence to using subset load balancing, but it will allow for a variety of other use cases in future.

Reply:
So a bunch of folks want to use subset name instead of subset-lb. No need to block this PR no the resolution of that question.

Reply:
This has nothing to do with subset lb. It is just metadata that becomes a NoOp if we don't allow people to use subset lb from destination rules. Envoy will not activate subset-lb without explicitly choosing subset lb (if that is your concern). I can guarantee that :).